### DISCUSSION TITLE
...i don't get it...

### DISCUSSION THREAD

**Discussion by madrooky** [Bing-su/adetailer#74] [Q&A]
Can someone, for a dummy like me, explain in a few words how this is supposed to be used? Please?
I got a1111, of course, and control net. So i get it i can create sort of masks, and then what?
For control net i kinda understand it, but not how i should use it without control net.

I do a lot of inpainting right now, and i would love to understand this thing and make it work. 

Thanks. :)

**Comment by spamnco** [comment]
Basically it takes care of face or hands inpainting for you.

1- Install the extension using this URL
2- RESTART your SD interface (important!)
3- select the new Adetailer tab and enable it
4- add your prompt, usually a simple <LORA> does the trick
5- enjoy!

If you want to take care of the hands, change the Adetailer model to hand_yolov8n although this part doesn't seem to work as well for me.

**Reply by metzo007** [reply]
this does not work. If I follow your instructions, no picture will be generatet because no mask was drawn.

**Reply by sussweb** [reply]
The masks are not visible. Think of it as "auto masking". Personally, I have not found out why I would want to use it on text2img. But when using img2img it's basically inpainting without the need to create a mask. E.g. you put an image of a person into img2img, enable adetailer, add a prompt like "face of [a different person]" in adetailer and the new face is inpainted.

**Reply by spamnco** [reply]
The use on txt2img is pretty obvious to me. When you generate a full body image in low resolution, the faces are always quite bad. Using this extension makes them automatically really good. Same for Loras which can limit the creativity of your first render. 
This extension doesn’t bring anything new to automatic1111 it just makes things easier and faster and it does it pretty well actually.

**Reply by Eleksar387** [reply]
Hm, need to play around with it some more. When I tested it in text2img and added a lora for a celebrity face the likeliness was much less than using the lora directly in the text2im prompt. It was actually barely recognizable. Maybe I did something wrong. Need to look into it again.

**Reply by spamnco** [reply]
It happens to me sometimes, I updated and fully restarted SD and now things are working fine again. You may have to increase the denoising strength.

**Reply by djwashout** [reply]
Can you lock in a set face to use on al your generations? one thing I dislike about the checkpoints is they tend to generate random faces throughout multiple images. This is of course once i figure out how to use this extension.. so far i cannot tell if its actually doing anything.

**Reply by djwashout** [reply]
I think I figured it out. For those of you who don't get it (like it didn't) think of it like this (someone correct me if I'm wrong) Its like a whole separate SD UI for marking, and for this use faces. Its quite nice because once I set it up I could tweak my face prompt inside ADetailer in such detail to get the face I want! This is great! Before SD wold just take the (Beautiful Face:1.3) prompt, and that was it. But with ADetailer I can can write a whole paragraph of a face prompt to get what I want! nice!

**Reply by andupotorac** [reply]
> s

It's much more than this. It automatically helps, during the process of image creation, to get your faces and hands right. So if you batch say 10 images, all of them will now have a better chance at the right face and proper hands.

I use it for my TI of my kid and at a first go images, 90% of them, don't do the face right. Or the hands. But after ADetailer goes through once more and paints those areas with the new prompt specific for hands or face, it's perfect. Increase the confidence value to 0.55 for best results.

An example:
![image](https://github.com/Bing-su/adetailer/assets/3269738/25523be2-dbc2-42b0-b1f0-cb05a2598ac5)

![image](https://github.com/Bing-su/adetailer/assets/3269738/8d2e6112-c258-4a80-903a-a282c2fbcaf0)

**Reply by djwashout** [reply]
How did you get the red highlight box around the face? I haven't got any option such as that. Its really like it doesn't work less I check it on or off. And even then I really cant tell if its ADetailer or the checkpoint. Are there any good tutorial on how to use this tool?

**Reply by andupotorac** [reply]
See reply from @spamnco below.

**Reply by CHollman82** [reply]
"It's much more than this. It automatically helps, during the process of image creation, to get your faces and hands right."

I have not noticed this. In txt2img it does not provide a "better chance" of getting good results, at least not in my experimenting. With faces they are just different, not better, and why would they be? With hands it is hit and miss, as hands are ALWAYS hit and miss... again it's just another random roll of the dice, there is a chance the 2nd result is better than the first.

I use it to quickly inpaint faces using img2img without having to manually mask as you would using inpaint. The downside is the entire image changes (as it does with img2img) instead of just the face.

**Comment by spamnco** [comment]
Go to the settings of afterdetailer and check the “save original render”
and “save detection zone” those are not the real name of the options but
you get the idea.
Then in your output folder you’ll see 3 images for each render. Makes
things much easier to understand.

**Reply by djwashout** [reply]
Humm... The only "settings" i have are drop down options for DETECTION, MASK PREPROCESSING, INPAINTING and under those sections i don't see an option to check the areas you listed. Is there another section I'm not seeing here?

**Reply by CHollman82** [reply]
He's talking about on the settings tab of A1111...

![Untitled](https://github.com/Bing-su/adetailer/assets/3136566/637715c0-52bf-4087-a152-f1f5a18fb991)

**Comment by spamnco** [comment]
It’s because the settings are under the settings tab of A1111

**Comment by ride5k** [comment]
having issues with using extension for img2img... usually i mask in inpaint mode, setup prompt/loras and away we go.  was hoping to automask using adetailer.  however, when starting in img2img inpaint mode with no manual mask, and adetailer enabled, face_yolov8n model, detection confidence 0.01 (for testing), generation fails and console dumps out ValueError: Coordinate 'right' is less than 'left' which is typical when a mask is not found.

not sure what i'm doing wrong... any help appreciated!

**Reply by spamnco** [reply]
Why don't you keep the default parameters? It's working fine for me. And don't use inpaint, just img2img, adetailer is taking care of the inpaint (that's the point).

**Reply by ride5k** [reply]
ok, using the img2img instead of inpaint is working.

however the bad news is that i could previously inpaint a much larger canvas since it was only editing a region, not the whole picture.  without using the inpainting module it is now re-rendering the entire picture, even if .  since i edit images with resolutions higher than my card can support, even if i dial denoise strength to zero in the "main" img2img dialog.

i found a potential work around by using the inpaint module, then painting a giant blob over the section i want to process, and then enabling adetailer.  the inpaint module does its samping thing where it cuts out the blob, then sends it over to adetailer.  faces are recognized properly and a "mask inside a mask" is drawn.  the images is rendered and reinserted into the final image whose dimensions are unchanged from the source.

**Reply by Technologicat** [reply]
@ride5k Thank you!

I also ran into this issue (ADetailer changes not saving in img2img inpaint mode). Face was detected and re-rendered, but the changes were not applied to the final output.

As you suggested, manually masking the area for inpaint fixed it. Doesn't need to be very accurate, just draw a huge blob as you suggested. Then the changes made by ADetailer apply correctly.

This kind of makes sense in retrospect. The whole point of inpaint mode is that it only ever applies changes to the masked area. It doesn't matter **how** the re-render takes place. The mask is applied to whatever the re-render produces.

Perhaps this could be documented more clearly.

**Comment by CHollman82** [comment]
I was a bit confused at first too because I was trying to use it with txt2img... and then I was like "why? it's just changing already good results".

Think of it instead as a way to automatically mask faces for inpainting so you don't have to draw the mask (do it in basic img2img though, not inpaint, inpaint will complain if you don't supply a mask and with this you don't need to).

You CAN use it with txt2img if you have a scene with many smaller faces, like a scene with a crowd, it seems to improve the faces in this case.

**Comment by spamnco** [comment]
Basically adetailer is an automated inpainting. One clever way to use it, is when using loras or lycoris, to inject those only at the inpainting step because they can limit the possibilities of some creative models.
For a full body render you can even use different layers: first body, then face and finally hands.
Remember that the actual size of a detail is linked to the quality, having 512x512 renders of a hand that is only in let's say 100x100 pixels makes the inpainting dramatically better.

**Comment by Rkkss** [comment]
How do you do it img2img mode?
I Send image to img2img, leave webui prompt & denoise at 0, set adetailer's inpainting to use seperate parameters and run the process. This way Img2img doesn't actually alter image and just use 2nd pass adetailer for inpainting.

It works, kind of...but only for certain samplers because most samplers require denoise value to be a non zero. Dunno this is optimal, can someone tells?
